utilities/ggd/lib/dag.go

```diff
@@ -0,0 +1,138 @@
+
+// String representation of the Import Graph.
+func (g ImportGraph) String() string {
+	var buffer bytes.Buffer
+	for pkg, importers := range g {
+		for importer := range importers {
+			buffer.WriteString(fmt.Sprintf("%s imported by %s\n", pkg, importer))
+		}
+	}
+	return buffer.String()
+}
+
+// ImportSet is a set of imports
+type ImportSet map[string]struct{}
+
+func NewImportGraph(ctx *build.Context, basePkg string) (ImportGraph, error) {
+	pkgs := buildutil.ExpandPatterns(ctx, []string{basePkg + "/..."})
+	g := make(ImportGraph, len(pkgs))
+
+	for pkg := range pkgs {
+		buildPkg, err := ctx.Import(pkg, "", 0)
+		if err != nil {
+			if _, ok := err.(*build.NoGoError); ok {
+				// The package doesn't contain any buildable go files. This can be caused, for
+				// example, when the only go files in a package contain build flags which are
+				// not present in our build context.
+				continue
+			}
+			return nil, fmt.Errorf("could not import package %v: %v", pkg, err)
+		}
+
+		visited := make(map[string]string)
+
+		// ensure we add the pkg to the graph
+		if _, ok := g[pkg]; !ok {
+			g[pkg] = make(ImportSet)
+		}
+
+		for _, path := range buildPkg.Imports {
+			g.maybeAddEdge(ctx, buildPkg, visited, pkg, path)
+		}
+		for _, path := range buildPkg.TestImports {
+			g.maybeAddEdge(ctx, buildPkg, visited, pkg, path)
+		}
+		for _, path := range buildPkg.XTestImports {
+			g.maybeAddEdge(ctx, buildPkg, visited, pkg, path)
+		}
+	}
+
+	return g, nil
+}
+
+func (g ImportGraph) maybeAddEdge(
+	ctx *build.Context, buildPkg *build.Package, visited map[string]string, pkg, path string,
+) {
+	if path == "C" {
+		// Not a real package.
+		return
+	}
+
+	importedPkg, ok := visited[path]
+	if !ok {
+		// It's okay for Import to return an error as not all packages that can be found in
+		// a package will necessarily be present. For example, packages imported only by test
+		// files in vendored packages will not be installed. In the case of an error, Import
+		// always returns a non-nil *Package. In the case of an error it will only contain
+		// partial information.
+		importBuildPkg, _ := ctx.Import(path, buildPkg.Dir, build.FindOnly)
+		if importBuildPkg != nil {
+			importedPkg = importBuildPkg.ImportPath
+		} else {
+			importedPkg = path
+		}
+		visited[path] = importedPkg
+		g.addEdge(importedPkg, pkg)
+	}
+}
+
+func (g ImportGraph) addEdge(from, to string) {
+	edges, ok := g[from]
+	if !ok {
+		edges = make(map[string]struct{})
+		g[from] = edges
+	}
+	edges[to] = struct{}{}
+}
+
+// Closure returns all the transitive closure of all packages reachable
+// by starting at the provided paths in the ImportGraph.
+func (g ImportGraph) Closure(paths ...string) ImportSet {
+	closure := make(ImportSet)
+	for _, p := range paths {
+		g.walk(p, closure)
+	}
+	return closure
+}
+
+func (g ImportGraph) walk(node string, visited ImportSet) {
+	if _, ok := visited[node]; ok {
+		return
+	}
+	visited[node] = struct{}{}
```

Comment:
I don't know if we can unconditionally add the package. For example, if we delete a package in our branch then it will be a changed package even though it no longer exists and `go test` will complain that it cant find it.

Reply:
Added `testcase7` for the case you're describing, here's the mini-diff: https://github.com/m3db/build-tools/pull/21/commits/f5b1fa0f81ac3c6d44ec1dc2c927bf226dad374b